### ISSUE TITLE
Load tensor from file in C++ fails 

### ISSUE THREAD

**Issue by jbcdnr** [pytorch/pytorch#20356]
## 🐛 Bug

I don't manage to import a tensor saved in PyTorch from C++. Any help on this is welcome 😄 

## To Reproduce

Save a tensor to a file in python.

```python
>>> import torch
>>> torch.save(torch.tensor([1., 2., 3.]), "tensor.pt")
>>> torch.load("tensor.pt")
tensor([1., 2., 3.])
```

Build this small C++ program

```cpp
#include <torch/torch.h>

int main() {
  torch::Tensor tensor;
  torch::load(tensor, "tensor.pt");
  std::cout << tensor << std::endl;
}
```

At runtime I get:

```
terminate called after throwing an instance of 'c10::Error'
  what():  [enforce fail at inline_container.cc:137] . PytorchStreamReader failed reading zip archive: failed finding central directory
frame #0: std::function<std::string ()>::operator()() const + 0x11 (0x7f1da515e691 in /___/libtorch/lib/libc10.so)
frame #1: c10::ThrowEnforceNotMet(char const*, int, char const*, std::string const&, void const*) + 0x49 (0x7f1da515e4a9 in /__/libtorch/lib/libc10.so)
frame #2: caffe2::serialize::PyTorchStreamReader::valid(char const*) + 0x6b (0x7f1d9c69735b in /__/libtorch/lib/libcaffe2.so)
frame #3: caffe2::serialize::PyTorchStreamReader::init() + 0x9d (0x7f1d9c69912d in /___/libtorch/lib/libcaffe2.so)
frame #4: caffe2::serialize::PyTorchStreamReader::PyTorchStreamReader(std::unique_ptr<caffe2::serialize::ReadAdapterInterface, std::default_delete<caffe2::serialize::ReadAdapterInterface> >) + 0x3b (0x7f1d9c69ab8b in /__/libtorch/lib/libcaffe2.so)
frame #5: <unknown function> + 0xa9b5bf (0x7f1da5e125bf in /__/libtorch/lib/libtorch.so.1)
frame #6: torch::jit::load(std::unique_ptr<caffe2::serialize::ReadAdapterInterface, std::default_delete<caffe2::serialize::ReadAdapterInterface> >, c10::optional<c10::Device>, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >&) + 0x9a (0x7f1da5e15d2a in /__/libtorch/lib/libtorch.so.1)
frame #7: torch::jit::load(std::string const&, c10::optional<c10::Device>, std::unordered_map<std::string, std::string, std::hash<std::string>, std::equal_to<std::string>, std::allocator<std::pair<std::string const, std::string> > >&) + 0x68 (0x7f1da5e15ec8 in /___/libtorch/lib/libtorch.so.1)
frame #8: torch::serialize::InputArchive::load_from(std::string const&, c10::optional<c10::Device>) + 0x38 (0x7f1da604c058 in /___/libtorch/lib/libtorch.so.1)
frame #9: void torch::load<at::Tensor, char const (&) [10]>(at::Tensor&, char const (&) [10]) + 0x86 (0x55a502f5fe97 in ./load)
frame #10: main + 0x37 (0x55a502f5fb51 in ./load)
frame #11: __libc_start_main + 0xe7 (0x7f1d9a338b97 in /lib/x86_64-linux-gnu/libc.so.6)
frame #12: _start + 0x2a (0x55a502f5f96a in ./load)

[1]    30788 abort      ./load
```
## Expected behavior

I would expect to be able to load a tensor from the PyTorch dump in the C++ program.

## Environment

PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 18.04.2 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: version 3.10.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.105
GPU models and configuration:
GPU 0: GeForce GTX TITAN X
GPU 1: GeForce GTX TITAN X
GPU 2: GeForce GTX TITAN X
GPU 3: GeForce GTX TITAN X
GPU 4: GeForce GTX TITAN X
GPU 5: GeForce GTX TITAN X
GPU 6: GeForce GTX TITAN X
GPU 7: GeForce GTX TITAN X

Nvidia driver version: 418.56
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] msgpack-numpy==0.4.3.2
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[pip] pytorch-nlp==0.4.1
[pip] torch==1.1.0
[pip] torchvision==0.2.2
[conda] blas                      1.0                         mkl
[conda] mkl                       2018.0.3                      1
[conda] mkl-service               1.1.2            py37h90e4bf4_5
[conda] mkl_fft                   1.0.6            py37h7dd41cf_0
[conda] mkl_random                1.0.1            py37h4414c95_1
[conda] pytorch                   1.1.0           py3.7_cuda9.0.176_cudnn7.5.1_0    pytorch
[conda] pytorch-nlp               0.4.1                    pypi_0    pypi
[conda] torchvision               0.2.2                      py_3    pytorch

## Additional context

I downloaded the C++ frontend from [https://download.pytorch.org/libtorch/nightly/cpu/libtorch-shared-with-deps-latest.zip](https://download.pytorch.org/libtorch/nightly/cpu/libtorch-shared-with-deps-latest.zip) yesterday.

**Comment by sakaia** [comment]
From Error message 
> PytorchStreamReader failed reading zip archive: failed finding central directory

As for ResNet18 model, it is zipped.
but tensor file is not zipped.

```
# file model.pt
rmodel.pt: Zip archive data
```

```
# file tensor.pt
tensor.pt: 8086 relocatable (Microsoft)
```

Ref.
  LOADING A PYTORCH MODEL IN C++
  https://pytorch.org/tutorials/advanced/cpp_export.html

**Comment by versine** [comment]
I have run into exactly the same problem trying to load a tensor into C++, and wondered if you ever found a solution?

**Comment by kwanUm** [comment]
+1, encountered this as well.

**Comment by minty99** [comment]
+1! I encountered this as well. Any updates on this?

**Comment by ThisIsIsaac** [comment]
@zou3519

**Comment by stiv-yakovenko** [comment]
Any news?

**Comment by driazati** [comment]
#25109 will fix this (though the API may change), #25502 needs to land first though

**Comment by hyc-xyz** [comment]
+1, any update?

**Comment by goksuguvendiren** [comment]
https://github.com/pytorch/pytorch/pull/25502 seems to have merged, are there any updates?

**Comment by driazati** [comment]
Sorry for the delay, #25502 only implements the serialization from C++ that can be loaded in Python, the path from Python -> C++ is still a WIP in #25109 which is blocked on some bigger serialization changes (see #26567 for details).

As a workaround in the meantime you can use `Module`s as a container to move values between C++ and Python:
```python
import torch

class Container(torch.nn.Module):
    def __init__(self, my_values):
        super().__init__()
        for key in my_values:
            setattr(self, key, my_values[key])

my_values = {
    'a': torch.ones(2, 2),
    'b': torch.ones(2, 2) + 10,
    'c': 'hello',
    'd': 6
}

# Save arbitrary values supported by TorchScript
# https://pytorch.org/docs/master/jit.html#supported-type
container = torch.jit.script(Container(my_values))
container.save("container.pt")
```

and load them in C++

```cpp
#include <torch/script.h>

#include <iostream>
#include <memory>

int main(int argc, const char *argv[]) {
  torch::jit::script::Module container = torch::jit::load("container.pt");

  // Load values by name
  torch::Tensor a = container.get_attribute("a").toTensor();
  std::cout << a << "\n";

  torch::Tensor b = container.get_attribute("b").toTensor();
  std::cout << b << "\n";

  std::string c = container.get_attribute("c").toStringRef();
  std::cout << c << "\n";

  int64_t d = container.get_attribute("d").toInt();
  std::cout << d << "\n";

  return 0;
}
```

**Comment by tsungruihon** [comment]
Any update please ? I tried to save a tensor from c++ libtorch and load it using python. And i got errors too.

**Comment by driazati** [comment]
It's still hidden behind a flag on the Python `torch.save` side that will eventually be flipped. Right now on nightly you can do this (copied from #22954):

Python -> C++
```python
import io
f = io.BytesIO()
torch.save(x, f, _use_new_zipfile_serialization=True)
# send f wherever
```

```cpp
// receive bytes in a `std::vector<char>`
std::vector<char> f = get_the_bytes();
torch::IValue x = torch::pickle_load(f);
```

and the reverse (C++ -> Python)

```cpp
torch::Tensor x = torch::ones({1, 2, 3});
std::vector<char> f = torch::pickle_save(x);
```

```python
x = torch.load(f)
```

**Comment by xiangpan-osu** [comment]
> It's still hidden behind a flag on the Python `torch.save` side that will eventually be flipped. Right now on nightly you can do this (copied from #22954):
> 
> Python -> C++
> 
> ```python
> import io
> f = io.BytesIO()
> torch.save(x, f, _use_new_zipfile_serialization=True)
> # send f wherever
> ```
> 
> ```c++
> // receive bytes in a `std::vector<char>`
> std::vector<char> f = get_the_bytes();
> torch::IValue x = torch::pickle_load(f);
> ```
> 
> and the reverse (C++ -> Python)
> 
> ```c++
> torch::Tensor x = torch::ones({1, 2, 3});
> std::vector<char> f = torch::pickle_save(x);
> ```
> 
> ```python
> x = torch.load(f)
> ```

For C++ -> Python case, if I wrote f into a file in C++ and run `x = torch.load(FILENAME)` in python it will encounter errors.

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/xiang.pan/Developments/python-test/env/local/lib/python2.7/site-packages/torch/serialization.py", line 426, in load
    return _load(f, map_location, pickle_module, **pickle_load_args)
  File "/home/xiang.pan/Developments/python-test/env/local/lib/python2.7/site-packages/torch/serialization.py", line 603, in _load
    magic_number = pickle_module.load(f, **pickle_load_args)
ValueError: unsupported pickle protocol: 10
```

Any ideas on how to do this? @driazati

**Comment by driazati** [comment]
Can you open a new issue with a full repro for your issue (and tag me in it)? For reference, this works fine for me on nightly PyTorch/libtorch:

```cpp
#include <torch/script.h>

#include <iostream>
#include <memory>

int main() {
  auto x = torch::ones({3, 3});
  auto bytes = torch::jit::pickle_save(x);
  std::ofstream fout("x.zip", std::ios::out | std::ios::binary);
  fout.write(bytes.data(), bytes.size());
  fout.close();
  return 0;
}
```

```python
import torch
torch.load("x.zip")
```

**Comment by xiangpan-osu** [comment]
> Can you open a new issue with a full repro for your issue (and tag me in it)? For reference, this works fine for me on nightly PyTorch/libtorch:
> 
> ```c++
> #include <torch/script.h>
> 
> #include <iostream>
> #include <memory>
> 
> int main() {
>   auto x = torch::ones({3, 3});
>   auto bytes = torch::jit::pickle_save(x);
>   std::ofstream fout("x.zip", std::ios::out | std::ios::binary);
>   fout.write(bytes.data(), bytes.size());
>   fout.close();
>   return 0;
> }
> ```
> 
> ```python
> import torch
> torch.load("x.zip")
> ```

Will do. Thanks for the pointers!

**Comment by veryluckyxyz** [comment]
> Sorry for the delay, #25502 only implements the serialization from C++ that can be loaded in Python, the path from Python -> C++ is still a WIP in #25109 which is blocked on some bigger serialization changes (see #26567 for details).
> 
> As a workaround in the meantime you can use `Module`s as a container to move values between C++ and Python:
> 
> ```python
> import torch
> 
> class Container(torch.nn.Module):
>     def __init__(self, my_values):
>         super().__init__()
>         for key in my_values:
>             setattr(self, key, my_values[key])
> 
> my_values = {
>     'a': torch.ones(2, 2),
>     'b': torch.ones(2, 2) + 10,
>     'c': 'hello',
>     'd': 6
> }
> 
> # Save arbitrary values supported by TorchScript
> # https://pytorch.org/docs/master/jit.html#supported-type
> container = torch.jit.script(Container(my_values))
> container.save("container.pt")
> ```
> 
> and load them in C++
> 
> ```c++
> #include <torch/script.h>
> 
> #include <iostream>
> #include <memory>
> 
> int main(int argc, const char *argv[]) {
>   torch::jit::script::Module container = torch::jit::load("container.pt");
> 
>   // Load values by name
>   torch::Tensor a = container.get_attribute("a").toTensor();
>   std::cout << a << "\n";
> 
>   torch::Tensor b = container.get_attribute("b").toTensor();
>   std::cout << b << "\n";
> 
>   std::string c = container.get_attribute("c").toStringRef();
>   std::cout << c << "\n";
> 
>   int64_t d = container.get_attribute("d").toInt();
>   std::cout << d << "\n";
> 
>   return 0;
> }
> ```

Hi,
Looks like [torch::jit::script::Module](https://pytorch.org/cppdocs/api/structtorch_1_1jit_1_1script_1_1_module.html#exhale-struct-structtorch-1-1jit-1-1script-1-1-module) no longer has `get_attribute` and I having a tough time understanding how to get named parameters using the API. Can you please suggest how to get named_parameters with the current API?

**Comment by driazati** [comment]
We're morphing the API to make it look more like [pybind11](https://github.com/pybind/pybind11), so now attribute accesses are via `attr` ([code](https://github.com/pytorch/pytorch/blob/ebe69236d192e9ccf5fd85b1652530b4487dd0c8/torch/csrc/jit/script/object.h#L57)). You should still be able to get `named_parameters` [directly](https://github.com/pytorch/pytorch/blob/b7c6277c53f82fb64234e95d06bf00be1bfc0002/torch/csrc/jit/script/module.h#L156). It looks like a bunch of this is missing from our C++ docs, we're working on cleaning it up in the next few weeks.

**Comment by veryluckyxyz** [comment]
Thank you for your response. i was able to get the values using `attr` but I still could not understand how to extract values from `slot_iterator_impl`. I will followup on it in [this](https://discuss.pytorch.org/t/loading-tensors-from-pytorch-to-torch-c/69261) 

Thank you again for your help!

**Comment by Edwardmark** [comment]
https://github.com/pytorch/pytorch/issues/36554
@yaatehr @oplatek @veryluckyxyz

**Comment by back2yes** [comment]
> 
> 
> > Can you open a new issue with a full repro for your issue (and tag me in it)? For reference, this works fine for me on nightly PyTorch/libtorch:
> > ```c++
> > #include <torch/script.h>
> > 
> > #include <iostream>
> > #include <memory>
> > 
> > int main() {
> >   auto x = torch::ones({3, 3});
> >   auto bytes = torch::jit::pickle_save(x);
> >   std::ofstream fout("x.zip", std::ios::out | std::ios::binary);
> >   fout.write(bytes.data(), bytes.size());
> >   fout.close();
> >   return 0;
> > }
> > ```
> > 
> > 
> > ```python
> > import torch
> > torch.load("x.zip")
> > ```
> 
> Will do. Thanks for the pointers!

For me in version 1.6, this does not work and throws an error:
`RuntimeError: [enforce fail at ..\caffe2\serialize\inline_container.cc:145] . PytorchStreamReader failed reading zip archive: failed finding central directory`

I managed to save tensors like this:

`in c++`
``` c++
auto x = torch::ones({ 300, 300 });
torch::save(x, "x.pt");
```

`in python`
``` python
data_fp = "path/to/x.pt"
model = torch.jit.load(data_fp)  # Yeah, libtorch saved our tensor in a model, as for version 1.6.
x = list(model.parameters())[0]
```

**Comment by gemfield** [comment]
Pytorch 1.6, tensor saved in c++ and load in python, get this warning:
```bash
/opt/conda/lib/python3.7/site-packages/torch/serialization.py:581: UserWarning: 'torch.load' received a zip file that looks like a TorchScript archive dispatching to 'torch.jit.load' (call 'torch.jit.load' directly to silence this warning)
  " silence this warning)", UserWarning)
```
And get return type <class 'torch.jit.RecursiveScriptModule'> after torch.load("xx.pt") in python (type Tensor should be expected).

BTW, @back2yes 's wa works.

**Comment by SpaceCowboy850** [comment]
This is so tantalizingly close to what I need to do.

My desire is that our pytorch team can save models like they normally do, like this:
`    torch.save(
        {
            'model': net.state_dict()
        },
        "./weights_net_10x.pth"
    )`

and then as they infer on those weights, we can take that exact same file and load it into our C++ production code.  For reasons beyond this scope, we aren't using libTorch to reconstruct the model.  While we have a method that CAN load weights via file conversion to another format, it would simplify logistics to just load the pth as it was saved in pytorch.

This is so close to what I want to do:
`        FILE* pFile = fopen("..\\..\\weights_net_10x.pth", "rb");

        fseek(pFile, 0, SEEK_END);
        long lSize = ftell(pFile);
        rewind(pFile);

        char* byte_buffer = new char[lSize];
        fread(byte_buffer, sizeof(char), lSize, pFile);

        std::vector<char> bytevec;
        bytevec.reserve(lSize);

        std::copy(&byte_buffer[0], &byte_buffer[lSize], std::back_inserter(bytevec));
        torch::IValue weights = torch::pickle_load(bytevec);`

but when I try this, I get a crash in unpickler.cpp, Unpickler::readInstruction, when trying to read the element in the stack after the string "model", it crashes as the index has tag of None, with payload of 0

`    case PickleOpCode::SETITEMS: {
      size_t start = marks_.back();
      marks_.pop_back();
      auto dict = stack_.at(start - 1).toGenericDict();
      for (size_t i = start; i < stack_.size(); i += 2) {
        dict.insert_or_assign(stack_[i], stack_[i + 1]);
      }
      stack_.erase(stack_.begin() + start, stack_.end());
    } break;`

All indices after that appear to be good, with layer name indices followed by the next index of what I assume are the parameterization of the layer.

Am I doing anything wrong here on either the CPP or Python side?  I can read in the weights with a jit.trace just fine, but from a workflow point of view, it would be really nice to _just_ read the weights as they are saved out by torch.save instead of going through a jit trace or torchScript.

**Comment by hyzhan** [comment]
> > Sorry for the delay, #25502 only implements the serialization from C++ that can be loaded in Python, the path from Python -> C++ is still a WIP in #25109 which is blocked on some bigger serialization changes (see #26567 for details).
> > As a workaround in the meantime you can use `Module`s as a container to move values between C++ and Python:
> > ```python
> > import torch
> > 
> > class Container(torch.nn.Module):
> >     def __init__(self, my_values):
> >         super().__init__()
> >         for key in my_values:
> >             setattr(self, key, my_values[key])
> > 
> > my_values = {
> >     'a': torch.ones(2, 2),
> >     'b': torch.ones(2, 2) + 10,
> >     'c': 'hello',
> >     'd': 6
> > }
> > 
> > # Save arbitrary values supported by TorchScript
> > # https://pytorch.org/docs/master/jit.html#supported-type
> > container = torch.jit.script(Container(my_values))
> > container.save("container.pt")
> > ```
> > 
> > 
> > and load them in C++
> > ```c++
> > #include <torch/script.h>
> > 
> > #include <iostream>
> > #include <memory>
> > 
> > int main(int argc, const char *argv[]) {
> >   torch::jit::script::Module container = torch::jit::load("container.pt");
> > 
> >   // Load values by name
> >   torch::Tensor a = container.get_attribute("a").toTensor();
> >   std::cout << a << "\n";
> > 
> >   torch::Tensor b = container.get_attribute("b").toTensor();
> >   std::cout << b << "\n";
> > 
> >   std::string c = container.get_attribute("c").toStringRef();
> >   std::cout << c << "\n";
> > 
> >   int64_t d = container.get_attribute("d").toInt();
> >   std::cout << d << "\n";
> > 
> >   return 0;
> > }
> > ```
> 
> Hi,
> Looks like [torch::jit::script::Module](https://pytorch.org/cppdocs/api/structtorch_1_1jit_1_1script_1_1_module.html#exhale-struct-structtorch-1-1jit-1-1script-1-1-module) no longer has `get_attribute` and I having a tough time understanding how to get named parameters using the API. Can you please suggest how to get named_parameters with the current API?

for libtorch 1.8, it use module.attr("XXX");

**Comment by buenos-dan** [comment]
> Sorry for the delay, #25502 only implements the serialization from C++ that can be loaded in Python, the path from Python -> C++ is still a WIP in #25109 which is blocked on some bigger serialization changes (see #26567 for details).
> 
> As a workaround in the meantime you can use `Module`s as a container to move values between C++ and Python:
> 
> ```python
> import torch
> 
> class Container(torch.nn.Module):
>     def __init__(self, my_values):
>         super().__init__()
>         for key in my_values:
>             setattr(self, key, my_values[key])
> 
> my_values = {
>     'a': torch.ones(2, 2),
>     'b': torch.ones(2, 2) + 10,
>     'c': 'hello',
>     'd': 6
> }
> 
> # Save arbitrary values supported by TorchScript
> # https://pytorch.org/docs/master/jit.html#supported-type
> container = torch.jit.script(Container(my_values))
> container.save("container.pt")
> ```
> 
> and load them in C++
> 
> ```c++
> #include <torch/script.h>
> 
> #include <iostream>
> #include <memory>
> 
> int main(int argc, const char *argv[]) {
>   torch::jit::script::Module container = torch::jit::load("container.pt");
> 
>   // Load values by name
>   torch::Tensor a = container.get_attribute("a").toTensor();
>   std::cout << a << "\n";
> 
>   torch::Tensor b = container.get_attribute("b").toTensor();
>   std::cout << b << "\n";
> 
>   std::string c = container.get_attribute("c").toStringRef();
>   std::cout << c << "\n";
> 
>   int64_t d = container.get_attribute("d").toInt();
>   std::cout << d << "\n";
> 
>   return 0;
> }
> ```

container has no member of get_attribute? why?

**Comment by Jackiexiao** [comment]
> It's still hidden behind a flag on the Python `torch.save` side that will eventually be flipped. Right now on nightly you can do this (copied from #22954):
> 
> Python -> C++
> ....

a full snippet from http://vra.github.io/2021/03/21/torch-tensor-python-to-cpp/

```python
import io

import torch


def save_tensor(device):
    my_tensor = torch.rand(3, 3).to(device);
    print("[python] my_tensor: ", my_tensor)
    f = io.BytesIO()
    torch.save(my_tensor, f, _use_new_zipfile_serialization=True)
    with open('my_tensor_%s.pt' % device, "wb") as out_f:
        # Copy the BytesIO stream to the output file
        out_f.write(f.getbuffer())


if __name__ == '__main__':
    save_tensor('cpu')
```

```cpp
#include <iostream>
#include <torch/torch.h>

std::vector<char> get_the_bytes(std::string filename) {
    std::ifstream input(filename, std::ios::binary);
    std::vector<char> bytes(
        (std::istreambuf_iterator<char>(input)),
        (std::istreambuf_iterator<char>()));

    input.close();
    return bytes;
}

int main()
{
    std::vector<char> f = get_the_bytes("my_tensor_cpu.pt");
    torch::IValue x = torch::pickle_load(f);
    torch::Tensor my_tensor = x.toTensor();
    std::cout << "[cpp] my_tensor: " << my_tensor << std::endl;

    return 0;
}
```

**Comment by Backlory** [comment]
So has this problem been solved now? Could we have a brief API in libtorch to do this?

**Comment by swilson314** [comment]
@driazati per your code above, this fails for me under windows with pytorch/libtorch 1.10.2. In response to the torch.load("x.zip") call I get the error: **PytorchStreamReader failed reading zip archive: failed finding central directory**. (I'm trying to move a state_dict from c++ to python so I can create onnx model. I get the same error when loading the state_dict.) Should I create a new issue?

```
#include <torch/script.h>
#include <iostream>
#include <memory>

int main() {
  auto x = torch::ones({3, 3});
  auto bytes = torch::jit::pickle_save(x);
  std::ofstream fout("x.zip", std::ios::out | std::ios::binary);
  fout.write(bytes.data(), bytes.size());
  fout.close();
  return 0;
}
```
```
import torch
torch.load("x.zip")
```
